Source/WebCore/layout/formattingContexts/inline/InlineItemsBuilder.cpp

```diff
@@ -711,6 +711,10 @@ void InlineItemsBuilder::breakAndComputeBidiLevels(InlineItemList& inlineItemLis
                 inlineItem.setBidiLevel(InlineItem::opaqueBidiLevel);
                 continue;
             }
+
+            if (inlineItem.isFloat())
+                continue;
+
```

Comment:
Maybe a more explicit check on whether we should _not_ assign InlineItem::opaqueBidiLevel to an inline box type of inline item would work here better. Something along the lines of

```
auto isContentfulInlineItem = [&] {
	if (auto* inlineTextItem = dynamicDowncast<InlineTextItem>(inlineItem))
		return !inlineTextItem->isWhitespace() || TextUtil::shouldPreserveSpacesAndTabs(inlineTextItem->layoutBox());
	return inlineItem.isAtomicInlineBox() || inlineItem.isLineBreak() || (inlineItem.isOpaque() && inlineItem.layoutBox().isOutOfFlowPositioned());
};
if (isContentfulInlineItem()) {
	// Mark the inline box stack with "content yes", when we come across a content type of inline item
	// so that we can mark the inline box as opaque and let the content drive visual ordering.
	inlineBoxContentFlagStack.fill(InlineBoxHasContent::Yes);
}
```

and mention in the commit message that <span><float box></span> and <span></span> are considered the same when it comes to finding visual ordering and that's why we need to keep a non-opaque bidi value on the [inline box start] so that it participates in reordering.

Reply:
Sadly this just partially addresses the issue here as we still end up constructing display boxes for all these empty inline boxes (while in non-bidi case we don't at all).